// @heroui/react/dist/components/switch/switch.js
"use client";
import { switchVariants } from '@heroui/styles';
import React__default, { createContext, useContext } from 'react';
import { Switch } from 'react-aria-components/Switch';
import { composeTwRenderProps, composeSlotClassName } from '../../utils/compose.js';
import { dom } from '../../utils/dom.js';
import { jsx, Fragment } from 'react/jsx-runtime';

const SwitchContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Switch Root
 * -----------------------------------------------------------------------------------------------*/

const SwitchRoot = ({
  children,
  className,
  size,
  ...props
}) => {
  const slots = React__default.useMemo(() => switchVariants({
    size
  }), [size]);
  return /*#__PURE__*/jsx(SwitchContext, {
    value: {
      slots
    },
    children: /*#__PURE__*/jsx(Switch, {
      "data-slot": "switch",
      ...props,
      className: composeTwRenderProps(className, slots.base()),
      children: values => /*#__PURE__*/jsx(Fragment, {
        children: typeof children === "function" ? children(values) : children
      })
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Switch Control
 * -----------------------------------------------------------------------------------------------*/

const SwitchControl = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(SwitchContext);
  return /*#__PURE__*/jsx(dom.span, {
    className: composeSlotClassName(slots?.control, className),
    "data-slot": "switch-control",
    ...props,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Switch Thumb
 * -----------------------------------------------------------------------------------------------*/

const SwitchThumb = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(SwitchContext);
  return /*#__PURE__*/jsx(dom.span, {
    className: composeSlotClassName(slots?.thumb, className),
    "data-slot": "switch-thumb",
    ...props,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Switch Icon
 * -----------------------------------------------------------------------------------------------*/

const SwitchIcon = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(SwitchContext);
  return /*#__PURE__*/jsx(dom.span, {
    className: composeSlotClassName(slots?.icon, className),
    "data-slot": "switch-icon",
    ...props,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Switch Content
 * -----------------------------------------------------------------------------------------------*/

const SwitchContent = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(SwitchContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.content, className),
    "data-slot": "switch-content",
    ...props,
    children: children
  });
};

export { SwitchContent, SwitchControl, SwitchIcon, SwitchRoot, SwitchThumb };
