// @heroui/react/dist/components/fieldset/fieldset.js
"use client";
import { fieldsetVariants } from '@heroui/styles';
import React__default, { createContext, useContext } from 'react';
import { composeSlotClassName } from '../../utils/compose.js';
import { dom } from '../../utils/dom.js';
import { jsx } from 'react/jsx-runtime';

const FieldsetContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Fieldset Root
 * -----------------------------------------------------------------------------------------------*/

const FieldsetRoot = ({
  className,
  ...props
}) => {
  const slots = React__default.useMemo(() => fieldsetVariants({}), []);
  return /*#__PURE__*/jsx(FieldsetContext, {
    value: {
      slots
    },
    children: /*#__PURE__*/jsx(dom.fieldset, {
      className: slots?.base({
        className
      }),
      "data-slot": "fieldset",
      ...props
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Fieldset Legend
 * -----------------------------------------------------------------------------------------------*/

const FieldsetLegend = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.legend, {
    className: composeSlotClassName(slots?.legend, className),
    "data-slot": "fieldset-legend",
    ...props
  });
};

/* -------------------------------------------------------------------------------------------------
 * Field Group
 * -----------------------------------------------------------------------------------------------*/

const FieldGroup = ({
  className,
  ...rest
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.fieldGroup, className),
    "data-slot": "fieldset-field-group",
    ...rest
  });
};

/* -------------------------------------------------------------------------------------------------
 * Field Actions
 * -----------------------------------------------------------------------------------------------*/

const FieldsetActions = ({
  children,
  className,
  ...rest
}) => {
  const {
    slots
  } = useContext(FieldsetContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.actions, className),
    "data-slot": "fieldset-actions",
    ...rest,
    children: children
  });
};

export { FieldGroup, FieldsetActions, FieldsetLegend, FieldsetRoot };
